### PR TITLE
doc: nrf9160_ug: add doc about nrf_modem integration

### PR DESCRIPTION
**Pull request by lemrey**
Add documentation about nrf_modem integration in NCS.

### CODE REVIEW

File: include/modem/nrf_modem_lib.rst [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+The library wrapper encapsulates the :c:func:`nrf_modem_init` and :c:func:`nrf_modem_shutdown` calls of the Modem library with :c:func:`nrf_modem_lib_init` and :c:func:`nrf_modem_lib_shutdown` calls, respectively.
+The library wrapper eases the task of initializing the Modem library by automatically passing the size and address of all the shared memory regions of the Modem library to the :c:func:`nrf_modem_init` call.
+
+The library wrapper can also initialize the Modem library during system initialization using :c:macro:`SYS_INIT`.
+The :option:`CONFIG_NRF_MODEM_LIB_SYS_INIT` Kconfig option can be used to control the initialization.
+Some libraries in |NCS|, such as the :ref:`at_cmd_readme` and the :ref:`lte_lc_readme` have similar options to initialize during system initialization and these options depend on the Modem library glue option.
+If your application performs an update of the nRF9160 modem firmware, you must disable this functionality to have full control on the initialization of the library.
+
+:ref:`partition_manager` is the component that reserves the RAM memory for the Modem library.
+For more information, see :ref:`partition_mgr_integration`.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:29]: Perhaps this could be moved closer to line 21 where we mention the shared memory. Also, instead of "reserves the RAM memory for the Modem library" we could specify "reserves the RAM memory for the shared memory regions used by the Modem library"

File: include/modem/nrf_modem_lib.rst [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+The library wrapper also coordinates the shutdown operation among different parts of the application that use the Modem library.
+This is done by the :c:func:`nrf_modem_lib_shutdown` function call, by invoking sleeping threads when the modem is being shut down.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:32]: "invoking" a thread is not correct, sleeping threads are "woken"

File: include/modem/nrf_modem_lib.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+When using the Modem library in |NCS|, the library should be initialized and shutdown using the :c:func:`nrf_modem_lib_init` and :c:func:`nrf_modem_lib_shutdown` function calls, respectively.
+
+Socket offloading
+=================
+
+Zephyr Socket API offers the :ref:`socket offloading functionality <zephyr:net_socket_offloading>` to redirect or *offload* function calls (for example, ``socket()`` and ``send()``) to BSD socket APIs.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:39]: I would rephrase this slightly because the BSD socket APIs are actually the APIs being offloaded here, so perhaps: ".. offload function calls to BSD socket APIs, for example, socket() and send()"

File: include/modem/nrf_modem_lib.rst [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+The glue layer utilizes this functionality to offload the socket API calls to the Modem library and thus eases the task of porting the networking code to the nRF9160 by providing a wrapper for Modem library's native socket API.
+Examples of such wrapper functions are :c:func:`nrf_socket` and :c:func:`nrf_send`.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:41]: Those are the library's native functions, which Zephyr's native socket calls e.g. `send()` are offloaded to. So this line is not correct, it is actually the Zephyr's native socket call `send()` that is wrapping ours `nrf_send` through offloading.

File: include/modem/nrf_modem_lib.rst [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+Note that the even if the socket offloading is disabled, Modem library's own socket APIs such as :c:func:`nrf_socket` and :c:func:`nrf_send` remain available.
+
+OS abstraction layer
+====================
+
+For functioning, the Modem library requires the implementation of an OS abstraction layer, which is an interface over the operating system functionalities such as interrupt setup, threads, and heap.
+The Modem library glue in |NCS| provides an implementation of the Modem library OS abstraction layer using |NCS| components.
+The OS abstraction layer is implemented in the :file:`nrfxlib\\nrf_modem\\include\\nrf_modem_os.c`.
+
+The behavior of the functions in the Modem library glue is dependent on the |NCS| components that are used in their implementation.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:62]: ".. of the functions in the Modem library _OS abstraction layer_ .."
Should be "OS abstraction layer" and not "glue.

File: include/modem/nrf_modem_lib.rst [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+.. _partition_mgr_integration:
+
+Partition manager integration
+=============================
+
+The Modem library, which runs on the application core, shares an area of RAM memory with the nRF9160 modem core.
+During the initialization, the Modem library accepts the boundaries of this area of RAM and configures the communication with the modem core accordingly.
+
+However, it is the responsibility of the application to reserve that RAM during linking, so that this memory area is not used for other purposes and remain dedicated for use by the Modem library.
+Therefore, the configuration parameters of the application decide the position and the size of the area of RAM dedicated to the Modem library.
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:75]: My bad, this sentence was not correct. I think we can remove it, the information that it's trying to convey is explained quite clearly in the sentence below.

File: include/modem/nrf_modem_lib.rst [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+* ``PM_NRF_MODEM_LIB_TX_ADDRESS`` - Address of the TX region
+* ``PM_NRF_MODEM_LIB_RX_ADDRESS`` - Address of the RX region
+* ``PM_NRF_MODEM_LIB_TRACE_ADDRESS`` - Address of the Trace region
+
+Partition manager also generates the following additional parameters:
+
+* ``PM_NRF_MODEM_LIB_CTRL_SIZE`` -Size of the Control region
+* ``CONFIG_PM_NRF_MODEM_LIB_TX_SIZE`` - Size of the TX region
+* ``CONFIG_PM_NRF_MODEM_LIB_RX_SIZE`` -Size of the RX region
+* ``CONFIG_PM_NRF_MODEM_LIB_TRACE_SIZE`` - Size of the Trace region
```
Comment (lemrey) [inline include/modem/nrf_modem_lib.rst:111]: `CONFIG_` prefix should be removed.

**Review by lemrey — COMMENTED** [review]
Thank you @umapraseeda, I have left a few comments.

File: doc/nrf/ug_nrf9160.rst [hunk 8]
```diff
@@ -121,6 +113,15 @@ Delta patches
   Delta patches are applied as firmware over-the-air (FOTA) upgrades.
   See :ref:`nrf9160_ug_fota` for more information.
 
+Modem library
```
[the remark this answers is not in the document]
Comment (umapraseeda) [reply doc/nrf/ug_nrf9160.rst:116]: Fixed

File: doc/nrf/ug_nrf9160.rst [hunk 9]
```diff
@@ -121,6 +113,15 @@ Delta patches
...
+-------------
+
+The |NCS| applications for the nRF9160 DK that communicate with the nRF9160 modem firmware must include the Modem library.
+The Modem library is released as an OS-independent binary library in the :ref:`nrfxlib:nrf_modem` repository and it is integrated into |NCS| via a glue layer, ``nrf_modem_lib``.
+
+The Modem library glue fulfills the integration requirements of the Modem library in |NCS|.
```
[the remark this answers is not in the document]
Comment (umapraseeda) [reply doc/nrf/ug_nrf9160.rst]: Agree

File: include/modem/nrf_modem_lib.rst [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+The Modem library integration layer handles the integration of the Modem library into |NCS|.
+The integration layer is constituted by the library wrapper and functionalities like socket offloading, OS abstraction, memory reservation by the Partition manager, and diagnostics.
+
+Library wrapper
```
[the remark this answers is not in the document]
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst:14]: Fixed

File: include/modem/nrf_modem_lib.rst (hunk 1 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Done

File: include/modem/nrf_modem_lib.rst (hunk 2 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Fixed

File: include/modem/nrf_modem_lib.rst (hunk 3 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Fixed

File: include/modem/nrf_modem_lib.rst (hunk 4 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Fixed

File: include/modem/nrf_modem_lib.rst (hunk 5 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Fixed

File: include/modem/nrf_modem_lib.rst (hunk 6 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Removed this.

File: include/modem/nrf_modem_lib.rst (hunk 7 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: Fixed

**Comment by lemrey** [comment]
@rlubos could you please have a look?

File: include/modem/nrf_modem_lib.rst [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+The size of the Control region is fixed.
+The Modem library exports the size value through :option:`CONFIG_NRF_MODEM_SHMEM_CTRL_SIZE`.
+This value is automatically passed by the integration layer to the library during the initialization through :c:func:`nrf_modem_lib_init`.
+
+When the application is built using CMake, the :ref:`partition_manager` automatically reads the Kconfig options of the integration layer.
+Partition manager decides about the placement of the regions in RAM and reserves memory according to the given size.
+As a result, the Partition manager generates the following parameters:
+
+* ``PM_NRF_MODEM_LIB_CTRL_ADDRESS`` -  Address of the Control region
```
Comment (rlubos) [inline include/modem/nrf_modem_lib.rst:99]:
```suggestion
* ``PM_NRF_MODEM_LIB_CTRL_ADDRESS`` - Address of the Control region
```
```suggestion
* ``PM_NRF_MODEM_LIB_CTRL_ADDRESS`` -  Address of the Control region
```

File: include/modem/nrf_modem_lib.rst [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+* ``PM_NRF_MODEM_LIB_TX_ADDRESS`` - Address of the TX region
+* ``PM_NRF_MODEM_LIB_RX_ADDRESS`` - Address of the RX region
+* ``PM_NRF_MODEM_LIB_TRACE_ADDRESS`` - Address of the Trace region
+
+Partition manager also generates the following additional parameters:
+
+* ``PM_NRF_MODEM_LIB_CTRL_SIZE`` -Size of the Control region
```
Comment (rlubos) [inline include/modem/nrf_modem_lib.rst:106]:
```suggestion
* ``PM_NRF_MODEM_LIB_CTRL_SIZE`` - Size of the Control region
```
```suggestion
* ``PM_NRF_MODEM_LIB_CTRL_SIZE`` -Size of the Control region
```

File: include/modem/nrf_modem_lib.rst [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+* ``PM_NRF_MODEM_LIB_TX_SIZE`` - Size of the TX region
+* ``PM_NRF_MODEM_LIB_RX_SIZE`` -Size of the RX region
```
Comment (rlubos) [inline include/modem/nrf_modem_lib.rst:108]:
```suggestion
* ``PM_NRF_MODEM_LIB_RX_SIZE`` - Size of the RX region
```
```suggestion
* ``PM_NRF_MODEM_LIB_RX_SIZE`` -Size of the RX region
```

**Review by rlubos — COMMENTED** [review]
Looks good, just a few nits spotted. Please squash the review fixes commit before merge.

File: include/modem/nrf_modem_lib.rst (hunk 11 above)
Comment (umapraseeda) [reply include/modem/nrf_modem_lib.rst]: fixed

> Looks good, just a few nits spotted. Please squash the review fixes commit before merge.

Thanks! Fixed these.

**Comment by lemrey** [comment]
Squashed, @rlubos @umapraseeda

**Review by rlubos — APPROVED** [review]

**Review by umapraseeda — APPROVED** [review]
